### PR TITLE
Updated _ingress.tpl to use ingress name as default for .service.name  if serviceName or .service.name is not defined.

### PR DESCRIPTION
**Pull request by drohnow**
This is a draft PR for the following shortcut: https://app.shortcut.com/replicated/story/71019/replicated-library-chart-allow-service-name-override-for-ingress.

Updated ../templates/.lib/_ingress.tpl

each .service.name field will be set with the following precedence:
- Default: $.ContextNames.ingress (ingress name) 
- If $values.serviceName exist (serviceName)
- If hosts[*].paths.service.name field exists (.service.name override)

Merely for review & feedback; have not bumped chart and added changes yet.

### CODE REVIEW

**Comment by diamonwiggins** [comment]
Makes sense to me. One thing I've been doing that Chris recommended is to add YAML snippets of what you've tested with. Just needs docs and this can be flipped to reviewable.

**Comment by drohnow** [comment]
Test:
   - Two ingress controllers defined (foo and bar) in the values.yaml.
   - No serviceName is defined for ingress controller foo.
   - The serviceName is defined for ingress controller bar.
   - Each has two hosts defined:
      - first host on each does not have hosts[*].paths.service.name manually defined
      - second host on each has a hosts[*].paths.service.name manually defined

    Expected Results: 
    - Ingress Controller foo:
       - First host (garage.rohnow.net) hosts[*].paths.service.name = foo (ingress controller name)
       - Second host (test.rohnow.net)hosts[*].paths.service.name = test-svc ((value of hosts[*].paths.service.name in values.yaml))
    - Ingress Controller bar:
       - First host (blog.rohnow.net) hosts[*].paths.service.name = webui (value of ingress.bar.serviceName in values.yaml)
       - Second host hosts[*].paths.service.name = corp-svc (value of hosts[*].paths.service.name in values.yaml)


image

image

image

File: charts/replicated-library/templates/lib/_ingress.tpl [hunk 1]
```diff
@@ -50,7 +50,10 @@ spec:
           {{- $service := $serviceName -}}
           {{- $port := 80 -}}
           {{- if .service -}}
-            {{- $service = default $service .service.name -}}
+            {{- $service = default $.ContextNames.ingress $values.serviceName -}}
```
Comment (chris-sanders) [inline charts/replicated-library/templates/lib/_ingress.tpl:53]: This defaults to the name of the ingress object, won't this need a prefix added to it for the default value to work correctly? Have you already tested it with the default prefix and this is taken care of somewhere else?

**Review by chris-sanders — COMMENTED** [review]
One question in-line that might be taken care of and I'm just not close enough to this part of the template to be sure of.

File: charts/replicated-library/templates/lib/_ingress.tpl (hunk 1 above)
Comment (diamonwiggins) [reply charts/replicated-library/templates/lib/_ingress.tpl:53]: This is super easy to forget at the moment.

**Comment by drohnow** [comment]
@diamonwiggins @chris-sanders  

Added prefix.

image

I verified that I can deploy the chart in the same namespace multiples times.

image

**Comment by drohnow** [comment]
Now that I have submitted the latest code, I am wondering if I should add a conditional to see if the service exists or even better that it is enabled?

File: charts/replicated-library/templates/lib/_ingress.tpl [hunk 2]
```diff
@@ -50,7 +50,11 @@ spec:
...
+              {{- if .service.name -}}
+             {{- $service = .service.name -}}
+            {{- end }}
+             {{- $service = printf "%s-%s" (include "replicated-library.names.prefix" $) $service -}}
```
Comment (diamonwiggins) [inline charts/replicated-library/templates/lib/_ingress.tpl:57]: Similar to https://github.com/replicatedhq/helm-charts/blob/main/charts/replicated-library/templates/lib/_pod_annotations.tpl#L40, you'll want to make sure you `trim` and `truncate` here.

**Comment by drohnow** [comment]
Added trim and truncate:

image

Used same values.yaml file as in comment above.

File: charts/replicated-library/templates/lib/_ingress.tpl (hunk 2 above)
Comment (drohnow) [reply charts/replicated-library/templates/lib/_ingress.tpl]: Thanks for the feedback.

**Review by chris-sanders — CHANGES_REQUESTED** [review]
I would still like to see us check that the provided service name is one from the chart (and enabled) before we add a prefix. If it's not both of those things we shouldn't be altering the user input.

It's possible for example, in an umbrella chart someone would be providing an explicit name and needs it used directly as provided.

**Comment by drohnow** [comment]
Added the following:
- check to see if service exists and is enabled; if so, add prefix.
- check to see if service exists and is NOT enabled; if true set service.name in accordance to default or serviceName condition; add prefix.
- if service.name value is not a service that is described within the chart, do not add prefix.

values.yaml snippet:
```
services:
 mongodb:
    enabled: true
    appName:
    fullNameOverride:
    type: ClusterIP
    clusterIP:
    externalTrafficPolicy:
    ipFamilyPolicy:
    ipFamilies: []
    annotations: {}
    labels:
    selector:
    ports:
      mongodb:
        enabled: true
        port: 27017
        protocol: TCP
        targetPort: 27017
        nodePort:

 webui:
    enabled: true
    appName:
    fullNameOverride:
    type: ClusterIP
    clusterIP:
    externalTrafficPolicy:
    ipFamilyPolicy:
    ipFamilies: []
    annotations: {}
    labels:
    selector:
    ports:
      webui:
        enabled: true
        port: 5000
        protocol: TCP
        targetPort: 5000

 disabled-svc:
   enabled: false
   appName:
   fullNameOverride:
   type: ClusterIP
   clusterIP:
   externalTrafficPolicy:
   ipFamilyPolicy:
   ipFamilies: []
   annotations: {}
   labels:
   selector:
   ports:
     webui:
       enabled: true
       port: 5000
       protocol: TCP
       targetPort: 5000 
  
  
ingresses:
  bar:
    enabled: true
    serviceName: webui
    fullNameOverride:
    annotations: {}
    labels: {}
    ingressClassName: contour
    hosts:
      - host: ai.rohnow.net
        paths:
          - path: /
            pathType: Prefix
            service:
              name: ai-svc # not an existing service in rLibChart; result should be: ai-svc
              port: 5000
      - host: corp.rohnow.net
        paths:
          - path: /
            pathType: Prefix
            service:
              name:      # service.name is null; result should be: release-name-garagersvp-webui (serviceName)
              port: 5000
      - host: corp2.rohnow.net
        paths:
          - path: /
            pathType: Prefix
            service:
              name: mongodb  # existing enabled service; result should be: release-name-garagersvp-mongodb (existing service)
              port: 5000  
    tls: []

  
  foo:
      enabled: true
      serviceName: 
      fullNameOverride:
      annotations: {}
      labels: {}
      ingressClassName: contour
      hosts:
        - host: blog.rohnow.net
          paths:
            - path: /
              pathType: Prefix
              service:
                name: mongodb # existing enabled service; result should be: release-name-garagersvp-mongodb (existing service)
                port: 5000
        - host: app1.rohnow.net
          paths:
            - path: /
              pathType: Prefix
              service:
                name:         # service.name is null; result should be: release-name-garagersvp-foo (default; name of ingress)
                port: 5000
        - host: foo2.rohnow.net
          paths:
            - path: /
              pathType: Prefix
              service:
                name: disabled-svc # an existing disabled service; result should be: release-name-garagersvp-foo (default; name of ingress)
                port: 5000
      tls: []

```

image

image


image

**Comment by drohnow** [comment]
code optimization; used same values.yaml and results were the same as above comment.

**Comment by chris-sanders** [comment]
> Added the following:
> 
> * check to see if service exists and is enabled; if so, add prefix.
> * check to see if service exists and is NOT enabled; if true set service.name in accordance to default or serviceName condition; add prefix.
> * if service.name value is not a service that is described within the chart, do not add prefix.
> 

This isn't what I was expecting. The first bullet point is correct. The 2nd bullet point is the one i think is incorrect. The prefix should only be added if the service is one we're installing. A disabled service is not getting installed and should be treated the same as if it were not there at all.

I have to run, I'll try to circle back later. I also think the defaults are not quite right. Defaults should not be set if we can't find a match. If nothing is provided and there are no services we can obviously use it should error.

**Comment by chris-sanders** [comment]
> @chris-sanders @diamonwiggins Adjusted code to error if service exists but is not enabled.
> 
> Give the following output:
> 
> ```
> disabled-svc is disabled, in order to utilize this service, it must be enabled.
> ```
> 
> note: "disabled-svc" is the service that is disabled in the rLibChart.
> 
> My thought is that it would be great to let the end user know which service is not enabled if they listed it in the .service.name field.

My concern with this is an end user (someone consuming a chart written with Library Chart) can turn off things but not remove them from the chart. As a contrived example if I want to use an Ingress for the `disabled-svc` which the chart also has implemented, but turned off. This current implementation would force that user to re-name their service, for this reason I propose if it's turned off it's treated exactly the same as if it were not defined at all.

**Comment by drohnow** [comment]
@chris-sanders

If I understand correctly, if the service is defined but disabled in the values.yaml file then we should treat the service.name as if it were not defined.  We cannot do that; manifests will render but it will not deploy.

Error: INSTALLATION FAILED: 1 error occurred:
	* Ingress.extensions "e-garagersvp-foo" is invalid: spec.rules[2].http.paths[0].backend.service.name: Required value

I like the check because if they have a service that is defined but disabled, it will tell them that they need to correct their values.yaml file (either remove the ingress host that has the disabled service defined if it is not needed or correct the service name they are using).

**Comment by drohnow** [comment]
> > Added the following:
> > 
> > * check to see if service exists and is enabled; if so, add prefix.
> > * check to see if service exists and is NOT enabled; if true set service.name in accordance to default or serviceName condition; add prefix.
> > * if service.name value is not a service that is described within the chart, do not add prefix.
> 
> This isn't what I was expecting. The first bullet point is correct. The 2nd bullet point is the one i think is incorrect. The prefix should only be added if the service is one we're installing. A disabled service is not getting installed and should be treated the same as if it were not there at all.
> 
> I have to run, I'll try to circle back later. I also think the defaults are not quite right. Defaults should not be set if we can't find a match. If nothing is provided and there are no services we can obviously use it should error.
> 
> Edit:
> 
> Alright so my thoughts on the defaults. This isn't actually as complicated as I thought it might be. The only modification I think I'd make here is that if you provide no service name _at all_ we should check to see if there is a service with the same name as the ingress and use that if it exists. That's kind of a last ditch effort, if it doesn't exist we should print a message that a service name is required.
> 
> An example of this is done with services themself and I'm just extending this to ingress. If you have an app `my-app` and a service `my-app` we assume that the service uses the app of the same name. This just extends that to assume if you also have an ingress `my-app` that this ingress goes with the service of the same name.
> 
> It isn't entirely uncommon for someone to do something like that and this affords them that option. It only really saves a couple of mapping variables and I wouldn't normally bother explaining to someone they can do this, just tell them to use the serviceName. But it's kind of nice if someone wants to do it this way.

What about allowing them to define a service that is outside of this chart? That is why I was checking for an existing service that is enabled and comparing it the service.name valu ; if it is an existing service that is enabled then add prefix; if it is not a service that is included in this chart then take it as what it is (do not add prefix). We still want this correct?

**Comment by drohnow** [comment]
Complete rework of conditionals; first blush, smoke tests look good.  I will update the PR in the morning to include tests.

**Comment by drohnow** [comment]
Test One:
- ingress name: webui (name of an existing service)
- serviceName: not defined
- Hosts: 
   - ai.rohnow.net:
      - host.paths.service.name: ai-svc (not an existing service)
      - expected results: ai-svc 
   - corp.rohnow.net: 
      - host.paths.service.name: mongodb (name of an existing & enabled service)
      - expected results: release-name-garagersvp-mongodb 
   - blog.rohnow.net: 
      - host.paths.service.name: disabled-svc (name of an existing & disabled service)
      - expected results: release-name-garagersvp-webui
    -kb.rohnow.net: 
      - host.paths.service.name:  null
      - expected results: release-name-garagersvp-webui
 
values.yaml:
```
services:
 mongodb:
    enabled: true
 ...

 webui:
    enabled: true
...

 disabled-svc:
   enabled: false
...
  
ingresses:
  webui:
    enabled: true
    serviceName: 
    fullNameOverride:
    annotations: {}
    labels: {}
    ingressClassName: contour
    hosts:
      - host: ai.rohnow.net
        paths:
          - path: /
            pathType: Prefix
            service:
              name: ai-svc          # not an existing service; expected result: ai-svc (named service.name)
              port: 5000
      - host: corp.rohnow.net
        paths:
          - path: /
            pathType: Prefix
            service:
              name: mongodb          # name of an existing & enabled service; expected result: release-name-garagersvp-mongodb (existing service)
              port: 5000
      - host: blog.rohnow.net
        paths:
          - path: /
            pathType: Prefix
            service:
              name:                  # null; expected result: release-name-garagersvp-webui (ingress name)
              port: 5000  
      - host: kb.rohnow.net
        paths:
          - path: /
            pathType: Prefix
            service:
              name: disabled-svc     # existing enabled service; expected result: release-name-garagersvp-webui (ingress name)
              port: 5000  
    tls: []

```
 
image

Test Two:
- ingress name: foo (name of not an existing service)
- serviceName: mongodb
- Hosts: 
   - ai.rohnow.net:
      - host.paths.service.name: ai-svc (not an existing service)
      - expected results: ai-svc 
   - corp.rohnow.net: 
      - host.paths.service.name: webui (name of an existing & enabled service)
      - expected results: release-name-garagersvp-webui 
   - blog.rohnow.net: 
      - host.paths.service.name: disabled-svc (name of an existing & disabled service)
      - expected results: release-name-garagersvp-mongodb
    -kb.rohnow.net: 
      - host.paths.service.name:  null
      - expected results: release-name-garagersvp-mongodb

values.yaml
```
services:
 mongodb:
    enabled: true
 ...
 webui:
    enabled: true
...
 disabled-svc:
   enabled: false
...

ingresses:
  foo:
    enabled: true
    serviceName: mongodb
    fullNameOverride:
    annotations: {}
    labels: {}
    ingressClassName: contour
    hosts:
      - host: ai.rohnow.net
        paths:
          - path: /
            pathType: Prefix
            service:
              name: ai-svc          # not an existing service; expected result: ai-svc (named service.name)
              port: 5000
      - host: corp.rohnow.net
        paths:
          - path: /
            pathType: Prefix
            service:
              name: webui           # name of an existing & enabled service; expected result: release-name-garagersvp-webui (existing servi
              port: 5000
      - host: blog.rohnow.net
        paths:
          - path: /
            pathType: Prefix
            service:
              name:                     # null; expected result: release-name-garagersvp-mongodb (defined service name)
              port: 5000  
      - host: kb.rohnow.net
        paths:
          - path: /
            pathType: Prefix
            service:
              name: disabled-svc     # existing enabled service; expected result: release-name-garagersvp-mongodb (defined service name)
              port: 5000  
    tls: []

```

image

Test Three:
- ingress name: foo (name of not an existing service)
- serviceName: not defined
- Hosts: 
   - ai.rohnow.net:
      - host.paths.service.name: ai-svc (not an existing service)
      - expected results: n/a
   - corp.rohnow.net: 
      - host.paths.service.name: webui (name of an existing & enabled service)
      - expected results: n/a
   - blog.rohnow.net: 
      - host.paths.service.name: disabled-svc (name of an existing & disabled service)
      - expected results: error -- service name is required for the ingress host
    -kb.rohnow.net: 
      - host.paths.service.name:  null
      - expected results: error -- a service name is required for the ingress host

values.yaml

```
services:
 mongodb:
    enabled: true
 ...

 webui:
    enabled: true
 ...

 disabled-svc:
   enabled: false
 ...
  
ingresses:
  foo:
    enabled: true
    serviceName: 
    fullNameOverride:
    annotations: {}
    labels: {}
    ingressClassName: contour
    hosts:
      - host: ai.rohnow.net
        paths:
          - path: /
            pathType: Prefix
            service:
              name: ai-svc          # n/a
              port: 5000
      - host: corp.rohnow.net
        paths:
          - path: /
            pathType: Prefix
            service:
              name: webui           # n/a
              port: 5000
      - host: blog.rohnow.net
        paths:
          - path: /
            pathType: Prefix
            service:
              name:                  # null; expected result: error -- a service name is required for the ingress host
              port: 5000  
      - host: kb.rohnow.net
        paths:
          - path: /
            pathType: Prefix
            service:
              name: disabled-svc     # existing enabled service; error -- a service name is required for the ingress host
              port: 5000  
    tls: []

```

image

**Comment by chris-sanders** [comment]
@drohnow can you double check if you have typos here?

Test 3
```
      - host: kb.rohnow.net
        paths:
          - path: /
            pathType: Prefix
            service:
              name: disabled-svc     # existing enabled service; error -- a service name is required for the ingress host
              port: 5000  
```

This is an existing disabled service, and I wouldn't expect an error. The user has provided a name, that name does not match a service we are installing, it should be used verbatim and assuming the user knows what they are doing. The error should only happen if they do not provide a name.

Test 2:
```
      - host: kb.rohnow.net
        paths:
          - path: /
            pathType: Prefix
            service:
              name: disabled-svc     # existing enabled service; expected result: release-name-garagersvp-mongodb (defined service name)
              port: 5000  
```

Here you've overridden the users input to a different service name. This again fails the one check that "A disabled service is treated like a service that isn't defined at all". In this example the user has provided a name, we should use that name with no prefix.

**Comment by drohnow** [comment]
Updated code; if service exists/disabled, and service.name value input..treat verbatim.
values.yaml
```
services:
 mongodb:
    enabled: true
 ...

 webui:
    enabled: true
 ...

 disabled-svc:
   enabled: false
 ...

ingresses:
  webui:
    enabled: true
    serviceName: 
    fullNameOverride:
    annotations: {}
    labels: {}
    ingressClassName: contour
    hosts:
      - host: ai.rohnow.net
        paths:
          - path: /
            pathType: Prefix
            service:
              name: ai-svc          # not an existing service; expected result: ai-svc (named service.name)
              port: 5000
      - host: corp.rohnow.net
        paths:
          - path: /
            pathType: Prefix
            service:
              name: mongodb          # name of an existing & enabled service; expected result: release-name-garagersvp
              port: 5000
      - host: blog.rohnow.net
        paths:
          - path: /
            pathType: Prefix
            service:
              name:                  # null; expected result: release-name-garagersvp-webui (ingress name)
              port: 5000  
      - host: kb.rohnow.net
        paths:
          - path: /
            pathType: Prefix
            service:
              name: disabled-svc     # existing disabled service; expected result: disabled-svc (named service.name)
              port: 5000  
    tls: []
```

image

**Comment by chris-sanders** [comment]
Great that looks good, it looks like Mongodb has an issue.

```
      - host: corp.rohnow.net
        paths:
          - path: /
            pathType: Prefix
            service:
              name: mongodb          # name of an existing & enabled service; expected result: release-name-garagersvp
              port: 5000
```
I believe you left off the name. Expected result: relase-name-garagersvp-mongodb

But yet the result i your test just shows "mongodb", given it's an existing enabled service it should have the prefix yes?

**Comment by drohnow** [comment]
values.yaml
```
services:
 mongodb:
    enabled: true
 ...
 webui:
    enabled: true
...
 disabled-svc:
   enabled: false
...

ingresses:
  webui:
    enabled: true
    serviceName:
    fullNameOverride:
    annotations: {}
    labels: {}
    ingressClassName: contour
    hosts:
      - host: ai.rohnow.net
        paths:
          - path: /
            pathType: Prefix
            service:
              name: ai-svc          # not an existing service; expected result: ai-svc (named service.name)
              port: 5000
      - host: corp.rohnow.net
        paths:
          - path: /
            pathType: Prefix
            service:
              name: mongodb          # name of an existing & enabled service; expected result: release-name-garagersvp-mongodb (existing service)
              port: 5000
      - host: blog.rohnow.net
        paths:
          - path: /
            pathType: Prefix
            service:
              name:                  # null; expected result: release-name-garagersvp-webui (ingress name)
              port: 5000  
      - host: kb.rohnow.net
        paths:
          - path: /
            pathType: Prefix
            service:
              name: disabled-svc     # existing enabled service; expected result: disabled-svc (named service.name)
              port: 5000  
    tls: []
```

image

File: charts/replicated-library/templates/lib/_ingress.tpl [hunk 3]
```diff
@@ -47,10 +47,30 @@ spec:
...
+          {{- end -}}
+          {{- if .service.name -}}
+              {{- $service = .service.name }}
+              {{- range $key, $val := $.Values.services }}
+                  {{- if and $val.enabled (eq $key $service) -}}
+                    {{- $service = printf "%s-%s" (include "replicated-library.names.prefix" $) $service | trunc 63 | trimAll "-"  -}}
+                  {{- end }}
+               {{- end }}
+            {{- end }}
+          {{- end }}
```
Comment (chris-sanders) [inline charts/replicated-library/templates/lib/_ingress.tpl:71]:
```suggestion
          {{- if .service.name }}
              {{- $service = .service.name }}
           {{- else if $values.serviceName }}
               {{- $service = $values.serviceName }}
           {{- else }}     
               {{- range $key, $val := $.Values.services }}
                   {{- if and $val.enabled (eq $key $.ContextNames.ingress) }}
                       {{- $service = $.ContextNames.ingress }}
                   {{- end }}
           {{- end }}
           {{- range $key, $val := $.Values.services }}
               {{- if and $val.enabled (eq $key $service) -}}
                   {{- $service = printf "%s-%s" (include "replicated-library.names.prefix" $) $service | trunc 63 | trimAll "-"  -}}
               {{- end }}
           {{- end }}
      {{- end }}
```


I _think_ this does the same thing all I've done is rearrange the logic for readability/maintainability:
* Use a single if / else set to make it clear which names are used first.
* Use a single check for the prefix to keep it DRY

I haven't tested this, I think the indention and end statements line up.

**Review by chris-sanders — CHANGES_REQUESTED** [review]
One in-line code organization request

**Comment by drohnow** [comment]
values.yaml:

```
services:
 mongodb:
    enabled: true
 ...

 webui:
    enabled: true
 ...

 disabled-svc:
   enabled: false
 ...
ingresses:
  webui:
    enabled: true
    serviceName:
    fullNameOverride:
    annotations: {}
    labels: {}
    ingressClassName: contour
    hosts:
      - host: ai.rohnow.net
        paths:
          - path: /
            pathType: Prefix
            service:
              name: ai-svc          # not an existing service; expected result: ai-svc (named service.name)
              port: 5000
      - host: corp.rohnow.net
        paths:
          - path: /
            pathType: Prefix
            service:
              name: mongodb          # name of an existing & enabled service; expected result: release-name-garagersvp-mongodb (existing service)
              port: 5000
      - host: blog.rohnow.net
        paths:
          - path: /
            pathType: Prefix
            service:
              name:                  # null; expected result: release-name-garagersvp-webui (ingress name)
              port: 5000  
      - host: kb.rohnow.net
        paths:
          - path: /
            pathType: Prefix
            service:
              name: disabled-svc     # existing enabled service; expected result: disabled-svc (named service.name)
              port: 5000  
    tls: []

```

image

**Review by chris-sanders — APPROVED** [review]
Looks like it's working to me +1

**Comment by drohnow** [comment]
All looks good!

values.yaml
 ```
services:
 mongodb:
    enabled: true
 ...

 webui:
    enabled: true
 ...

 disabled-svc:
   enabled: false
 ...
ingresses:
  webui:
    enabled: true
    serviceName:
    fullNameOverride:
    annotations: {}
    labels: {}
    ingressClassName: contour
    hosts:
      - host: ai.rohnow.net
        paths:
          - path: /
            pathType: Prefix
            service:
              name: ai-svc          # not an existing service; expected result: ai-svc (named service.name)
              port: 5000
      - host: corp.rohnow.net
        paths:
          - path: /
            pathType: Prefix
            service:
              name: mongodb          # name of an existing & enabled service; expected result: release-name-garagersvp-mongodb (existing service)
              port: 5000
      - host: blog.rohnow.net
        paths:
          - path: /
            pathType: Prefix
            service:
              name:                  # null; expected result: release-name-garagersvp-webui (ingress name)
              port: 5000  
      - host: kb.rohnow.net
        paths:
          - path: /
            pathType: Prefix
            service:
              name: disabled-svc     # existing enabled service; expected result: disabled-svc (named service.name)
              port: 5000  
    tls: []

```

image

**Comment by chris-sanders** [comment]
+1 looks good

**Pull request merged**